dir = "/home/UTHSCSA/huelgasmoral/Documents"
SAMPLES= ["105T", "203T", "235T", "266T", "367T", "369T", '444T', "472T"]

rule bwa_mem:
    input:
        reads=["{dir}/WES/NGS-WES-RNAseq-UT_Austin_Nov2013/JA13476_WES/{sample}.1.fastq.gz",
        "{dir}/WES/NGS-WES-RNAseq-UT_Austin_Nov2013/JA13476_WES/{sample}.2.fastq.gz"],
    output:
        "{dir}/WES/Analysis/mapped/{sample}.bam",
    log:
        "{dir}/WES/Analysis/logs/bwa_mem/{sample}.log",
    params:
        index="{dir}/Reference/Homo_sapiens_assembly38.fasta",
        extra="",
        sorting="none",  # Can be 'none', 'samtools' or 'picard'.
        sort_order="coordinate",  # Can be 'queryname' or 'coordinate'.
        sort_extra="",  # Extra args for samtools/picard.
        tmp_dir="/home/UTHSCSA/huelgasmoral/tmp",  # Path to temp dir. (optional)
    threads: 8
    wrapper:
        "0.80.2/bio/bwa/mem"

rule samtools_sort:
    input:
        "{dir}WES/Analysis/mapped/{sample}.bam"
    output:
        "{dir}WES/Analysis/mapped/{sample}.sorted.bam"
    params:
        extra = "-m 4G",
        tmp_dir = "/home/UTHSCSA/huelgasmoral/tmp"
    threads:  # Samtools takes additional threads through its option -@
        8     # This value - 1 will be sent to -@.
    wrapper:
        "0.80.2/bio/samtools/sort"

rule replace_rg:
    input:
        "{dir}WES/Analysis/mapped/{sample}.sorted.bam"
    output:
        "{dir}WES/Analysis/fixed-rg/{sample}.bam"
    log:
        "{dir}WES/Analysis/logs/picard/replace_rg/{sample}.log"
    params:
        "RGLB=lib1 RGPL=illumina RGPU={sample} RGSM={sample}"
    # optional specification of memory usage of the JVM that snakemake will respect with global
    # resource restrictions (https://snakemake.readthedocs.io/en/latest/snakefiles/rules.html#resources)
    # and which can be used to request RAM during cluster job submission as `{resources.mem_mb}`:
    # https://snakemake.readthedocs.io/en/latest/executing/cluster.html#job-properties
    resources:
        mem_mb=1024
    wrapper:
        "0.80.2/bio/picard/addorreplacereadgroups"

rule samtools_index:
    input:
        "{dir}WES/Analysis/fixed-rg/{sample}.bam"
    output:
        "{dir}WES/Analysis/fixed-rg/{sample}.bam.bai"
    #log:
    #    "logs/samtools_index/{sample}.log"
    params:
        "" # optional params string
    threads:  # Samtools takes additional threads through its option -@
        4     # This value - 1 will be sent to -@
    wrapper:
        "0.80.2/bio/samtools/index"

rule haplotype_caller:
    input:
        # single or list of bam files
        bam="{dir}WES/Analysis/fixed-rg/{sample}.bam",
        ref="{dir}Reference/Homo_sapiens_assembly38.fasta"
        # known="dbsnp.vcf"  # optional
    output:
        gvcf="{dir}WES/Analysis/calls/{sample}.g.vcf",
#   bam="{sample}.assemb_haplo.bam",
    log:
        "{dir}WES/Analysis/logs/gatk/haplotypecaller/{sample}.log"
    params:
        extra="--tmp-dir /home/UTHSCSA/huelgasmoral/tmp",  # optional
        java_opts="", # optional
    resources:
        mem_mb=6000
    wrapper:
        "0.80.2/bio/gatk/haplotypecaller"
